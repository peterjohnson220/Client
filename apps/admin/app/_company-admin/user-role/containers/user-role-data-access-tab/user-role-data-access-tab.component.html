<div *ngIf="currentRole$ | async;else no_role">
  <div *ngIf="!(currentRole$ | async)?.IsSystemRole;else system_role_restriction_message" class="row">
    <div class="col-xl-10 col-12">
      <div class="alert alert-info">
        Create user roles to permit or restrict users from accessing specific data. Users assigned to a role cannot
        access any data that does not meet the criteria defined below.<br />
        Select <span class="text-danger">Is equal to</span> to allow users access to only data defined in the field and
        select <span class="text-danger">Is not equal to</span> to allow
        users access to everything except data defined in the field. If there are no restrictions set, users will have
        access to all data for that entity.
      </div>
    </div>
    <div *ngFor="let dt of dataTypes$ | async" class="col-xl-10 col-12">
      <div *ngIf="dt.Name !== 'SurveysCompanySurveys'" class="d-flex align-items-center pb-2">
        <div class="flex-grow-1 mr-4">
          <b>{{dt.Name}}</b>
        </div>
        <div class="action-buttons">
          <button type="button" *ngIf="dt.DataFields.length > 1 && !dt.AddAndDeleteDisabled" (click)="addDataRestriction(dt)"
            class="btn btn-secondary action-buttons">
            <fa-icon icon="plus"></fa-icon>
          </button>
        </div>
      </div>
      <div *ngFor="let rdr of roleDataRestrictions$ | async | dataTypeFilter:dt;trackBy: trackByFn;"
        class="d-flex align-items-center pb-2">
        <div class="flex-grow-1 mr-4">
          <pf-data-field-filter [dataType]="dt" [roleDataRestriction]="rdr"
            (roleDataRestrictionChanged)="roleDataRestrictionChanged(rdr, $event.property, $event.value)">
          </pf-data-field-filter>
        </div>
        <div class="action-buttons">
          <button type="button" *ngIf="dt.DataFields.length > 1 && !dt.AddAndDeleteDisabled" (click)="removeDataRestriction(rdr)"
            class="btn btn-secondary">
            <fa-icon icon="trash"></fa-icon>
          </button>
        </div>
      </div>
      <ng-container *ngIf="!dt.HideDivider">
        <hr />
      </ng-container>
    </div>
  </div>
</div>

<ng-template #no_role>
  <div>
    <h4>Please select a role from the role list.</h4>
  </div>
</ng-template>

<ng-template #system_role_restriction_message>
  <div>
    <h4>No data restrictions can be added to the {{(currentRole$ | async)?.RoleName}} role. If you would like to add
      data restrictions, please create a new user role.</h4>
  </div>
</ng-template>
