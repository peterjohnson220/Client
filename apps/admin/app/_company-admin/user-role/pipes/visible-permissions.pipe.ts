import { Pipe, PipeTransform } from '@angular/core';
import { SystemPermission } from 'libs/models/security/roles/system-permission.model';

@Pipe({name: 'visiblePermission'})
export class VisiblePermissionsPipe implements PipeTransform {
  transform(permissions: SystemPermission[]) {
    if (!permissions) {
      return;
    }
    return permissions.filter(p => p.UiVisible);
  }
}
