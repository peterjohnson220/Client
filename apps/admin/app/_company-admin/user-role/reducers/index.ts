import { createFeatureSelector, createSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromUserRoleViewReducer from './user-role-view.reducer';
import * as fromUserRoleUserTabReducer from './user-role-users-tab.reducer';
import * as fromUserRoleFunctionTabReducer from './user-role-functions-tab.reducer';
import * as fromDataAccessTabReducer from './role-data-access-tab.reducer';

// Feature area state
export interface CompanyAdminUserRolesState {
  userRoleView: fromUserRoleViewReducer.IUserRoleState;
  userRoleUserTab: fromUserRoleUserTabReducer.State;
  userRoleFunctionTab: fromUserRoleFunctionTabReducer.State;
  dataAccessTab: fromDataAccessTabReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  companyAdminUserRoles: CompanyAdminUserRolesState;
}

// Feature area reducers
export const reducers = {
  userRoleView: fromUserRoleViewReducer.reducer,
  userRoleUserTab: fromUserRoleUserTabReducer.reducer,
  userRoleFunctionTab: fromUserRoleFunctionTabReducer.reducer,
  dataAccessTab: fromDataAccessTabReducer.reducer,
};

// Select Feature Area
export const selectCompanyAdminUserRolesState =
  createFeatureSelector<CompanyAdminUserRolesState>('companyAdminUserRoles');

// View Selectors
export const selectUserRoleState =
  createSelector(selectCompanyAdminUserRolesState, (state: CompanyAdminUserRolesState) => state.userRoleView);

export const selectUserRoleFunctionTabState =
  createSelector(selectCompanyAdminUserRolesState, (state: CompanyAdminUserRolesState) => state.userRoleFunctionTab);

export const selectRoleDataAccessTabState =
  createSelector(selectCompanyAdminUserRolesState, (state: CompanyAdminUserRolesState) => state.dataAccessTab);

export const selectUserRoleUserTabState =
  createSelector(selectCompanyAdminUserRolesState, (state: CompanyAdminUserRolesState) => state.userRoleUserTab);

// Select User Role
export const getUserRoleViewState = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getUserRoleViewState
);

export const getUserRoleCurrentTabState = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getUserRoleCurrentTab
);

export const getCurrentUserRole = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getCurrentUserRole
);

export const getCompanyRoles = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getCompanyRoles
);

export const getAddCompanyRoleModalIsOpen = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleModalIsOpen
);

export const getAddCompanyRoleForm = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleForm
);

export const getAddCompanyRoleError = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleError
);

export const getRoleApiResponse = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getRoleApiResponse
);

export const getSaveButtonDisabled = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getSaveButtonDisabled
);

// Functions tab
export const getFunctionTabPermissions = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getFunctionTabPermissions
);

export const getFunctionTabPendingChanges = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getFunctionTabPendingChanges
);

export const getCurrentCheckedPermissionIds = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getCheckedPermissions
);

export const getCheckboxesDisabled = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getCheckboxesDisabled
);

// Users tab
export const getUsersAndRoles = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersAndRoles
);

export const getUsersInActiveRole = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersInActiveRole
);

export const getUsersNotInActiveRole = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersNotInActiveRole
);

export const getUsersTabHasPendingChanges = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersTabHasPendingChanges
);

export const getUserIdsToSave = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUserIdsToSave
);

// Data Access Tab
export const getDataTypes = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getDataTypes
);

export const getRoleDataRestrictions = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getRoleDataRestrictions
);

export const getRoleDataRestrictionsForSave = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getRoleDataRestrictionsForSave
);

export const getDataAccessTabPendingChanges = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getDataAccessTabHasPendingChanges
);
