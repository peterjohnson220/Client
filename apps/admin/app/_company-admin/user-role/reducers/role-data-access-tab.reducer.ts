import * as fromDataAccessTabActions from '../actions/data-access-tab.action';
import { DataField, DataType, RoleDataRestriction } from 'libs/models/security/roles';
import * as cloneDeep from 'lodash.clonedeep';
import { DataFieldTypes } from '../constants/data-field-type.constants';

export interface State {
  currentRoleId: number;
  dataTypes: DataType[];
  dataFields: DataField[];
  currentRoleDataRestrictions: RoleDataRestriction[];
  currentRoleDataRestrictionsUnchanged: RoleDataRestriction[];
}

export const initialState: State = {
  currentRoleId: 0,
  dataTypes: [],
  dataFields: [],
  currentRoleDataRestrictions: [],
  currentRoleDataRestrictionsUnchanged: []
};

export function reducer(state = initialState, action: fromDataAccessTabActions.DataAccessTabAction): State {
  switch (action.type) {
    case fromDataAccessTabActions.LOADED_DATA_TYPES: {
      let dataFields = [];
      (action.payload).forEach(dt => {
        dataFields = dataFields.concat(dt.DataFields);
      });
      return {
        ...state,
        dataTypes: action.payload,
        dataFields: dataFields
      };
    }
    case fromDataAccessTabActions.CANCEL_ROLE_DATA_RESTRICTIONS_CHANGES:
      return {
        ...state,
        currentRoleDataRestrictions: cloneDeep(state.currentRoleDataRestrictionsUnchanged
          .map(dr => ({ ...dr, Id: new Date().getUTCMilliseconds() + dr.Id })))
      };
    case fromDataAccessTabActions.ADD_NEW_DATA_RESTRICTION: {
      const dataFieldId = state.dataFields.find(dt => dt.DataTypeId === action.dataType.Id).Id;
      const dataRestriction = {
        Id: new Date().getUTCMilliseconds(),
        RoleId: state.currentRoleId,
        DataFieldId: dataFieldId,
        DataValue: null,
        DataConditionIsEqual: undefined
      };
      return Object.assign({}, state, {
        currentRoleDataRestrictions: [...state.currentRoleDataRestrictions, dataRestriction]
      });
    }
    case fromDataAccessTabActions.REMOVE_DATA_RESTRICTION: {
      const dataRestrictions = [...state.currentRoleDataRestrictions];
      const index = state.currentRoleDataRestrictions.indexOf(action.dataRestriction);
      dataRestrictions.splice(index, 1);
      return {
        ...state,
        currentRoleDataRestrictions: dataRestrictions
      };
    }
    case fromDataAccessTabActions.UPDATE_SINGLE_DATA_RESTRICTION: {
      const dataRestrictions = [...state.currentRoleDataRestrictions];
      const index = state.currentRoleDataRestrictions.indexOf(action.dataRestriction);
      const dataRestriction = { ...state.currentRoleDataRestrictions[index] };
      dataRestriction[action.property] = action.value;
      dataRestrictions[index] = dataRestriction;
      return {
        ...state,
        currentRoleDataRestrictions: dataRestrictions
      };
    }
    case fromDataAccessTabActions.UPDATE_CURRENT_ROLE: {
      if (!action.payload) {
        return {
          ...state,
          currentRoleDataRestrictions: initialState.currentRoleDataRestrictions,
          currentRoleDataRestrictionsUnchanged: initialState.currentRoleDataRestrictionsUnchanged,
          currentRoleId: initialState.currentRoleId
        };
      }
      const withBlankRows = createBlankDataRestrictionIfNeeded(state.dataTypes, action.payload.DataRestrictions);
      const dataRestrictions = convertRoleDataRestrictionForUI(state.dataFields, withBlankRows);
      return {
        ...state,
        currentRoleDataRestrictions: dataRestrictions,
        currentRoleDataRestrictionsUnchanged: dataRestrictions.map(dr => ({ ...dr, Id: new Date().getUTCMilliseconds() + dr.Id })),
        currentRoleId: action.payload.RoleId
      };
    }
    default: {
      return state;
    }
  }
}
export const getDataTypes = (state: State) => state.dataTypes;

export const getRoleDataRestrictions = (state: State) => state.currentRoleDataRestrictions;

export const getRoleDataRestrictionsForSave = (state: State) =>
  convertRoleDataRestrictionForSave(state.dataFields, state.currentRoleDataRestrictions, state.currentRoleId);

export const getDataAccessTabHasPendingChanges = (state: State) => {
  let changed = false;
  const stringCurrentRoleDataRestrictionsUnchanged = getRoleDataAccessString(state.currentRoleDataRestrictionsUnchanged);
  getRoleDataAccessString(state.currentRoleDataRestrictions).forEach(f => {
    if (!stringCurrentRoleDataRestrictionsUnchanged.find(uch => uch === f)) {
      changed = true;
      return;
    }
  });
  return changed;
};

function getRoleDataAccessString(roleDataPermissions: RoleDataRestriction[]) {
  return roleDataPermissions ? roleDataPermissions
    .map(r => `${r.DataFieldId}_${r.DataConditionIsEqual}_${r.DataValue}`) : [];
}

function createBlankDataRestrictionIfNeeded(dataTypes: DataType[], roleDataRestrictions: RoleDataRestriction[]) {
  const dataRestrictions = [];
  dataTypes.forEach(dt => {
    const rd = roleDataRestrictions.find(r => dt.DataFields.map(m => m.Id).indexOf(r.DataFieldId) > -1);
    if (!rd) {
      dataRestrictions.push({ Id: new Date().getUTCMilliseconds(), DataFieldId: dt.DataFields[0].Id, DataConditionIsEqual: true });
    }
  });
  return dataRestrictions.concat(roleDataRestrictions);
}

function convertRoleDataRestrictionForUI(dataFields: DataField[], roleDataRestrictions: RoleDataRestriction[]) {
  let dataRestrictions = [];

  dataFields.forEach(df => {
    const tempRd = roleDataRestrictions.filter(dr => dr.DataFieldId === df.Id);
    if (tempRd && tempRd.length !== 0) {
      switch (df.FieldType) {
        case DataFieldTypes.MULTISELECT:
          dataRestrictions.push({ ...tempRd[0], Id: new Date().getUTCMilliseconds(), DataValue: tempRd.map(t => t.DataValue) });
          break;
        default:
          dataRestrictions = dataRestrictions.concat(tempRd.map(dr => ({ ...dr })));
          break;
      }
    }
  });

  return dataRestrictions;
}


function convertRoleDataRestrictionForSave(dataFields: DataField[], dataRestrictions: RoleDataRestriction[], roleId: number) {
  const newRoleDataRestrictions = [];
  dataRestrictions.filter(f => f.DataFieldId).forEach(r => {
    if (dataFields.find(df => df.Id === r.DataFieldId).FieldType === DataFieldTypes.MULTISELECT) {
      r.DataValue.forEach(dv => {
        newRoleDataRestrictions.push({
          DataFieldId: r.DataFieldId,
          RoleId: roleId,
          DataConditionIsEqual: r.DataConditionIsEqual,
          DataValue: dv
        });

      });
    } else {
      newRoleDataRestrictions.push({
        DataFieldId: r.DataFieldId,
        RoleId: roleId,
        DataConditionIsEqual: r.DataConditionIsEqual,
        DataValue: r.DataValue
      });
    }
  });
  return newRoleDataRestrictions;
}
