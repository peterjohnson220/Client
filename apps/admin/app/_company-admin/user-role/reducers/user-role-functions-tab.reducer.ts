import * as cloneDeep from 'lodash.clonedeep';

import { SystemPermission } from 'libs/models/security/roles';
import { arraysEqual } from 'libs/core/functions';

import * as fromUserRoleFunctionTabActions from '../actions/user-role-functions-tab.action';

export interface State {
  currentPermissions: SystemPermission[];
  currentPermissionsDefaultStatus: SystemPermission[];
  disableCheckboxes: boolean;
}

export const initialState: State = {
  currentPermissions: undefined,
  currentPermissionsDefaultStatus: undefined,
  disableCheckboxes: false
};

export function reducer(state = initialState, action: fromUserRoleFunctionTabActions.FunctionTabActions): State {
  switch (action.type) {
    case fromUserRoleFunctionTabActions.UPDATE_CURRENT_ROLE_FUNCTION_TAB: {
      return action.payload ? {
        ...state,
        currentPermissions: action.payload.Permissions,
        currentPermissionsDefaultStatus: action.payload.Permissions,
        disableCheckboxes: action.payload.IsSystemRole
      } : {
          ...state,
          currentPermissions: initialState.currentPermissions,
          currentPermissionsDefaultStatus: initialState.currentPermissionsDefaultStatus
        };
    }
    case fromUserRoleFunctionTabActions.GRANT_DENY_PERMISSIONS: {
      const currPermission = action.payload;
      const newRolePermissions = cloneDeep(state.currentPermissions);

      if (currPermission.IsParent) {
        newRolePermissions.forEach(p => {
          if (p.SystemPermissionId === currPermission.SystemPermissionId) {
            p.IsChecked = !currPermission.IsChecked;
            if (p.ChildPermission) {
              p.ChildPermission.forEach(cp => cp.IsChecked = !currPermission.IsChecked);
            }
            return;
          }
        });
      } else {
        newRolePermissions.forEach(p => {
          if (p.ChildPermission) {
            const index = p.ChildPermission.findIndex(f => f.SystemPermissionId === currPermission.SystemPermissionId);
            if (index > -1) {
              p.ChildPermission[index].IsChecked = !currPermission.IsChecked;
            }
          }
        });
      }

      return {
        ...state,
        currentPermissions: newRolePermissions
      };
    }
    case fromUserRoleFunctionTabActions.CANCEL_PERMISSION_CHANGES: {
      return {
        ...state,
        currentPermissions: state.currentPermissionsDefaultStatus
      };
    }
    default: {
      return state;
    }
  }
}

export const getUserRoleViewState = (state: State) => state;
export const getFunctionTabPermissions = (state: State) => state.currentPermissions;
export const getCheckedPermissions = (state: State) => getCheckedPermissionIds(state.currentPermissions);
export const getCheckboxesDisabled = (state: State) => state.disableCheckboxes;
export const getFunctionTabPendingChanges = (state: State) =>
  !(arraysEqual(getCheckedPermissionIds(state.currentPermissionsDefaultStatus), getCheckedPermissionIds(state.currentPermissions)));

function getCheckedPermissionIds(permissionCollection: SystemPermission[]): number[] {
  if (!permissionCollection) {
    return [];
  }
  const checkedIds: number[] = [];
  permissionCollection.forEach(p => {
    if (p.IsChecked) {
      checkedIds.push(p.SystemPermissionId);
      if (p.IsParent) {
        p.ChildPermission.forEach(cp => {
          if (cp.IsChecked) {
            checkedIds.push(cp.SystemPermissionId);
          }
        });
      }
    }
  });

  return checkedIds;
}
