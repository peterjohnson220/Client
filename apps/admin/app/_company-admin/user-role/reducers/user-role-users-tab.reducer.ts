import cloneDeep from 'lodash/cloneDeep';

import { arraySortByString, SortDirection } from 'libs/core/functions';
import { UserAndRoleModel, UserAssignedRole } from 'libs/models/security/roles';

import * as fromUserRoleUserTabActions from '../actions/user-role-users-tab.action';

export interface State {
  users: UserAndRoleModel[];
  originalUsers: UserAndRoleModel[];
  activeRole: UserAssignedRole;
  filterTerm: string;
}

export const initialState: State = {
  users: [],
  originalUsers: [],
  activeRole: null,
  filterTerm: ''
};

export function reducer(state = initialState, action: fromUserRoleUserTabActions.UserTabActions): State {
  switch (action.type) {
    case fromUserRoleUserTabActions.UPDATE_CURRENT_USER_ROLE_USER_TAB: {
      const newRole = action.payload;
      return {
        ...state,
        activeRole: newRole,
      };
    }
    case fromUserRoleUserTabActions.GET_USERS_AND_ROLES: {
      return {
        ...state,
      };
    }
    case fromUserRoleUserTabActions.GET_USERS_AND_ROLES_SUCCESS: {
      return {
        ...state,
        users: Object.assign([], action.payload),
        originalUsers: Object.assign([], action.payload)
      };
    }
    case fromUserRoleUserTabActions.ADD_USER_TO_ROLE: {
      let updatedUsers = Object.assign([], state.users);
      const userToUpdate = cloneDeep(updatedUsers.find(u => u.UserId === action.payload.UserId));
      userToUpdate.CurrentRoleId = state.activeRole.RoleId;
      userToUpdate.RoleName = state.activeRole.RoleName;
      userToUpdate.Dirty = true;

      updatedUsers = updatedUsers.filter(u => u.UserId !== userToUpdate.UserId);
      updatedUsers.push(userToUpdate);

      return {
        ...state,
        users: updatedUsers
      };
    }
    case fromUserRoleUserTabActions.FILTER_USERS_COLLECTION: {
      return {
        ...state,
        filterTerm: action.payload.toString().toLowerCase(),
      };
    }
    case fromUserRoleUserTabActions.CANCEL_CHANGES: {
      if (!state.activeRole) {
        return state;
      }

      return {
        ...state,
        users: Object.assign([], state.originalUsers)
      };
    }
    default: {
      return state;
    }
  }
}

export const getUserRoleUserTabState = (state: State) => state;
export const getUsersAndRoles = (state: State) => state.users;
export const getUsersInActiveRole = (state: State) => getUsersInRole(state);
export const getUsersNotInActiveRole = (state: State) => getUsersNotInRole(state);
export const getUsersTabHasPendingChanges = (state: State) => state.users.some(u => u.Dirty);
export const getUserIdsToSave = (state: State) => state.users.filter(u => u.Dirty).map(u => u.UserId);

function getUsersInRole(state: State) {
  return state.activeRole ? state.users
    .filter(user => user.CurrentRoleId === state.activeRole.RoleId)
    .sort((a, b) => arraySortByString(a.LastName, b.LastName, SortDirection.Ascending)) : [];
}

function getUsersNotInRole(state: State) {
  return state.activeRole ? state.users
    .filter(user =>
      (user.FirstName.toLowerCase().indexOf(state.filterTerm) > -1
        || user.LastName.toLowerCase().indexOf(state.filterTerm) > -1
        || user.UserId.toString().toLowerCase().indexOf(state.filterTerm) > -1
      ) && user.CurrentRoleId !== state.activeRole.RoleId
    )
    .sort((a, b) => arraySortByString(a.LastName, b.LastName, SortDirection.Ascending)) : [];
}
