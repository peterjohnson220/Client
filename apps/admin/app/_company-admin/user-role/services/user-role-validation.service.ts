import { Injectable } from '@angular/core';
import { FormControl } from '@angular/forms';

@Injectable()
export class UserRoleValidationService {

  NAME_LENGTH = { min: 1, max: 255 };

  constructor() { }

  validateSpecialChars(control: FormControl) {
    // allow letters, numbers, commas, periods, apostrophes, and hyphens
    const nameRegex = /^[a-zA-Z0-9 ,.'-]+$/i;

    let name = control.value;
    if (name) { name = name.trim(); }

    return nameRegex.test(name) ? null : { 'invalidCharacter': true };
  }
}
