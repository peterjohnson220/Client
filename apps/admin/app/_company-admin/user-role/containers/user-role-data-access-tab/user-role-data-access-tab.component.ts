import { Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { DataType, RoleDataRestriction, UserAssignedRole } from 'libs/models/security/roles';
import * as fromRoleDataTabActions from '../../actions/data-access-tab.action';
import * as userRoleReducer from '../../reducers';

@Component({
  selector: 'pf-user-role-data-access-tab',
  templateUrl: './user-role-data-access-tab.component.html',
  styleUrls: ['./user-role-data-access-tab.component.scss']
})
export class UserRoleDataAccessTabComponent {
  dataTypes$: Observable<DataType[]>;
  roleDataRestrictions$: Observable<RoleDataRestriction[]>;
  currentRole$: Observable<UserAssignedRole>;

  constructor(public store: Store<userRoleReducer.State>) {
    this.dataTypes$ = this.store.select(userRoleReducer.getDataTypes);
    this.roleDataRestrictions$ = this.store.select(userRoleReducer.getRoleDataRestrictions);
    this.currentRole$ = this.store.select(userRoleReducer.getCurrentUserRole);
  }

  roleDataRestrictionChanged(dataRestriction: RoleDataRestriction, property: string, value: any) {
    this.store.dispatch(new fromRoleDataTabActions.UpdateSingleDataRestriction(dataRestriction, property, value));
  }

  removeDataRestriction(dataRestriction: RoleDataRestriction) {
    this.store.dispatch(new fromRoleDataTabActions.RemoveDataRestriction(dataRestriction));
  }

  addDataRestriction(dataType: DataType) {
    this.store.dispatch(new fromRoleDataTabActions.AddNewDataRestriction(dataType));
  }
  trackByFn(index, item: RoleDataRestriction) {
    return item.Id;
  }
}
