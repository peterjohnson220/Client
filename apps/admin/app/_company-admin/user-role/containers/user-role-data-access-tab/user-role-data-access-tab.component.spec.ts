import { TestBed, waitForAsync } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { combineReducers, Store, StoreModule } from '@ngrx/store';

import { of } from 'rxjs';

import * as fromRootState from 'libs/state/state';
import { DataTypeFilterPipe } from 'libs/core/pipes';
import { getMockDataTypes } from 'libs/models/security/roles/data-type.model';
import { getMockRoleDataRestrictions } from 'libs/models/security/roles/role-data-restriction.model';
import { generateMockSystemDefinedUserAssignedRole, generateMockUserAssignedRole } from 'libs/models/security/roles/user-assigned-roles.model';

import { UserRoleDataAccessTabComponent } from './user-role-data-access-tab.component';
import * as fromUserRoleViewReducer from '../../reducers';
import * as fromDataAccessTabActions from '../../reducers';
import { DataAccessService } from '../../services';

describe('UserRoleDataAccessTabComponent', () => {
  let fixture, component;
  let store: Store<fromUserRoleViewReducer.State>;
  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          companyAdminUserRoles: combineReducers(fromDataAccessTabActions.reducers)
        }),
      ],
      providers: [DataAccessService],
      declarations: [UserRoleDataAccessTabComponent, DataTypeFilterPipe],
      schemas: [NO_ERRORS_SCHEMA]
    });
    store = TestBed.inject(Store);

    spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(UserRoleDataAccessTabComponent);
    component = fixture.componentInstance;  // to access properties and methods
    component.roleDataRestrictions$ = of(getMockRoleDataRestrictions());
    component.dataTypes$ = of(getMockDataTypes());
    component.currentRole$ = of(generateMockUserAssignedRole());
  }));

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should show data access tab on company role', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should display message on system role', () => {
    component.currentRole$ = of(generateMockSystemDefinedUserAssignedRole());
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
