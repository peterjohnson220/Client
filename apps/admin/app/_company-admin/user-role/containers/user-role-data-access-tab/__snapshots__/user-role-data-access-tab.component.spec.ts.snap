// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRoleDataAccessTabComponent should display message on system role 1`] = `
<pf-user-role-data-access-tab
  __ngContext__={[Function LRootView]}
  currentRole$={[Function Observable]}
  dataTypes$={[Function Observable]}
  roleDataRestrictions$={[Function Observable]}
  store={[Function Store]}
>
  <div>
    <div>
      <h4>
        No data restrictions can be added to the Test Role role. If you would like to add data restrictions, please create a new user role.
      </h4>
    </div>
  </div>
</pf-user-role-data-access-tab>
`;

exports[`UserRoleDataAccessTabComponent should show data access tab on company role 1`] = `
<pf-user-role-data-access-tab
  __ngContext__={[Function LRootView]}
  currentRole$={[Function Observable]}
  dataTypes$={[Function Observable]}
  roleDataRestrictions$={[Function Observable]}
  store={[Function Store]}
>
  <div>
    <div
      class="row"
    >
      <div
        class="col-xl-10 col-12"
      >
        <div
          class="alert alert-info"
        >
           Create user roles to permit or restrict users from accessing specific data. Users assigned to a role cannot access any data that does not meet the criteria defined below.
          <br />
           Select 
          <span
            class="text-danger"
          >
            Is equal to
          </span>
           to allow users access to only data defined in the field and select 
          <span
            class="text-danger"
          >
            Is not equal to
          </span>
           to allow users access to everything except data defined in the field. If there are no restrictions set, users will have access to all data for that entity. 
        </div>
      </div>
      <div
        class="col-xl-10 col-12"
      >
        <div
          class="d-flex align-items-center pb-2"
        >
          <div
            class="flex-grow-1 mr-4"
          >
            <b>
              Test Data type
            </b>
          </div>
          <div
            class="action-buttons"
          >
          </div>
        </div>
        <div
          class="d-flex align-items-center pb-2"
        >
          <div
            class="flex-grow-1 mr-4"
          >
            <pf-data-field-filter />
          </div>
          <div
            class="action-buttons"
          >
          </div>
        </div>
        <hr />
      </div>
    </div>
  </div>
</pf-user-role-data-access-tab>
`;
